man/unerf.1
.Dd July 24, 2018
.Dt UNERF 1
.Os
.Sh NAME
.Nm unerf
.Nd BioWare ERF (.erf, .mod, .nwm, .sav) archive extractor
.Sh SYNOPSIS
.Nm unerf
.Op Ar options
.Ar command
.Ar archive
.Op Ar
.Sh DESCRIPTION
.Nm
extracts BioWare ERF archives, found in many BioWare games as files
with the extension .erf, .mod, .nwm or .sav.
Moreover, in some games, a .rim file might be an ERF instead of a RIM.
.Pp
There's several different versions of ERFs.
This tool supports the version V1.0, V1.1, V2.0, V2.2 and V3.0.
.Pp
Supported features:
.Bl -bullet -compact
.It
Archives with stripped filenames. The name hash is searched for in
a big lookup table.
If it is not found, the file is listed and
extracted as a hexadecimal representation of the hash
.It
zlib-compressed files are supported and transparently decompressed
.It
Blowfish-encrypted archives are supported and are, provided you know
the password, transparently decrypted
.It
Archives that contain full paths are supported.
You can either strip the directory and extract the plain filenames,
or substitute the directory separator with a
.Ql =
.It
Archives with numerical extension IDs belonging to different games
can be aliased accordingly.
.El
.Pp
Unsupported features:
.Bl -bullet -compact
.It
XOR-encrypted archives
.It
Creating directories when extracting archives with full paths.
.Pp
Please also note that by stripped the directories of files in these
archives, you might overwrite already extracted files
.El
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl Fl nwn2
Alias file types according to
.Em Neverwinter Nights 2
rules.
.Pp
.Em Neverwinter Nights 2
reuses a few file extension IDs differently than other BioWare games.
To correctly read Neverwinter Nights 2 ERF archives, use this flag.
.It Fl Fl jade
Alias file types according to
.Em Jade Empire
rules.
.Pp
.Em Jade Empire
reuses a few file extension IDs differently than other BioWare games.
To correctly read Jade Empire ERF archives, use this flag.
.It Fl Fl pass Ar hex
Password to decrypt this archive.
This password needs to be specified in hexadecimal notation,
even if it's an ASCII string.
For example,
.Dq 4CF223AB
would be a valid password string.
.Dq Foobar
would not.
.It Fl Fl nwm Ar file
Calculate the MD5 of this NWM file to complement the decryption key
of a HAK file for a Neverwinter Nights premium module.
.El
.Bl -tag -width xxxx -compact
.It Ar command
.Bl -tag -width xx -compact
.It Cm i
Display meta-information about the archive
.It Cm l
List filesystem contents (without directories)
.It Cm v
List filesystem contents with full paths
.It Cm e
Extract files to current directory, stripping directories
.It Cm x
Extract files to current directory with full path, including directories.
.El
.It Ar archive
The ERF archive to read.
.It Ar file
One or more files to extract.
If none are given, the whole archive is extracted.
.El
.Sh EXAMPLES
View meta-information of the archive
.Pa Shadowlords1.mod :
.Pp
.Dl $ unerf i Shadowlords1.mod
.Pp
List all files contained in the archive
.Pa Chapter1.nwm :
.Pp
.Dl $ unerf l Chapter1.nwm
.Pp
List all files, with full path, in the archive
.Pa areas.erf :
.Pp
.Dl $ unerf v areas.erf
.Pp
Extract all files from the archive
.Pa foo.mod :
.Pp
.Dl $ unerf e foo.mod
.Pp
Extract the file
.Pa area1.are
from the archive
.Pa foo.mod :
.Pp
.Dl $ unerf e foo.mod area1.are
.Pp
Extract all files, with full path, in the archive
.Pa areas.erf :
.Pp
.Dl $ unerf x areas.erf
.Pp
Extract the file
.Pa areas\earea1.are
from the archive
.Pa areas.erf :
.Pp
.Dl $ unerf x areas.erf areas\e\earea1.are
.Sh SEE ALSO
.Xr erf 1 ,
.Xr fixpremiumgff 1 ,
.Xr unherf 1 ,
.Xr unrim 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/unerf.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to extract ERF (.erf, .mod, .nwm, .sav) archives.
 */

#include <cstring>
#include <cstdio>

#include <vector>
#include <set>

#include "src/version/version.h"

#include "src/common/ustring.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/md5.h"
#include "src/common/cli.h"

#include "src/aurora/util.h"
#include "src/aurora/erffile.h"

#include "src/archives/util.h"

#include "src/util.h"

enum Command {
	kCommandNone        = -1,
	kCommandInfo        =  0,
	kCommandList            ,
	kCommandListVerbose     ,
	kCommandExtract         ,
	kCommandExtractDir      ,
	kCommandMAX
};

const char *kCommandChar[kCommandMAX] = { "i", "l", "v", "e", "x" };

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Command &command, Common::UString &archive, std::set<Common::UString> &files,
                      Aurora::GameID &game, std::vector<byte> &password);

bool parsePassword(const Common::UString &arg, std::vector<byte> &password);
bool readNWMMD5   (const Common::UString &arg, std::vector<byte> &password);

void displayInfo(Aurora::ERFFile &erf);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Aurora::GameID game = Aurora::kGameIDUnknown;

		int returnValue = 1;
		Command command = kCommandNone;
		Common::UString archive;
		std::set<Common::UString> files;
		std::vector<byte> password;

		if (!parseCommandLine(args, returnValue, command, archive, files, game, password))
			return returnValue;

		Aurora::ERFFile erf(new Common::ReadFile(archive), password);
		files = Archives::fixPathSeparator(files);

		if      (command == kCommandInfo)
			displayInfo(erf);
		else if (command == kCommandList)
			Archives::listFiles(erf, game, false);
		else if (command == kCommandListVerbose)
			Archives::listFiles(erf, game, true);
		else if (command == kCommandExtract)
			Archives::extractFiles(erf, game, false, files);
		else if (command == kCommandExtractDir)
			Archives::extractFiles(erf, game, true, files);

	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parsePassword(const Common::UString &arg, std::vector<byte> &password) {
	const size_t length = arg.size();

	password.clear();
	password.reserve(length / 2);

	size_t i = 0;
	byte c = 0x00;
	for (Common::UString::iterator s = arg.begin(); s != arg.end(); ++s, i++) {
		byte d = 0;

		if      (*s >= '0' && *s <= '9')
			d = *s - '0';
		else if (*s >= 'a' && *s <= 'f')
			d = *s - 'a' + 10;
		else if (*s >= 'A' && *s <= 'F')
			d = *s - 'A' + 10;
		else
			throw Common::Exception("0x%08X is not a valid hex digit", (uint) *s);

		if ((i % 2) == 1) {
			c |= d;

			password.push_back(c);

			c = 0x00;
		} else
			c |= d << 4;
	}
	return true;
}

bool readNWMMD5(const Common::UString &arg, std::vector<byte> &password) {
	Common::ReadFile keyFile(arg);

	Common::hashMD5(keyFile, password);
	return true;
}

namespace Common {
namespace CLI {
template<>
int ValGetter<Command &>::get(const std::vector<Common::UString> &args, int i, int) {
	_val = kCommandNone;
	for (int j = 0; j < kCommandMAX; j++) {
		if (!strcmp(args[i].c_str(), kCommandChar[j])) {
			_val = (Command) j;
			return 0;
		}
	}
	return -1;
}
}
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Command &command, Common::UString &archive, std::set<Common::UString> &files,
                      Aurora::GameID &game, std::vector<byte> &password) {

	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::Callback;
	using Common::CLI::ValGetter;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Aurora::GameID;

	NoOption cmdOpt(false, new ValGetter<Command &>(command, "command"));
	NoOption archiveOpt(false, new ValGetter<Common::UString &>(archive, "archive"));
	NoOption filesOpt(true, new ValGetter<std::set<Common::UString> &>(files, "files[...]"));
	Parser parser(argv[0], "BioWare ERF (.erf, .mod, .nwm, .sav) archive extractor",
	              "Commands:\n"
	              "  i          Display meta-information\n"
	              "  l          List files (stripping directories)\n"
	              "  v          List files verbosely (with directories)\n"
	              "  e          Extract files to current directory, stripping directories\n"
	              "  x          Extract files to current directory, creating subdirectories\n",
	              returnValue,
	              makeEndArgs(&cmdOpt, &archiveOpt, &filesOpt));

	parser.addSpace();
	parser.addOption("nwn2", "Alias file types according to Neverwinter Nights 2 rules",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("jade", "Alias file types according to Jade Empire rules",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
	parser.addSpace();
	parser.addOption("pass", "Decryption password, if required, in hex notation",
	                 kContinueParsing,
	                 new Callback<std::vector<byte> &>("hex", parsePassword, password));
	parser.addOption("nwn",
	                 "Neverwinter Nights premium module file(for decrypting their HAK file)",
	                 kContinueParsing,
	                 new Callback<std::vector<byte> &>("file", readNWMMD5, password));

	return parser.process(argv);
}

void displayInfo(Aurora::ERFFile &erf) {
	std::printf("Version: %s\n", Common::debugTag(erf.getVersion()).c_str());
	std::printf("Build Year: %d\n", erf.getBuildYear());
	std::printf("Build Day: %d\n", erf.getBuildDay());
	std::printf("Number of files: %s\n", Common::composeString(erf.getResources().size()).c_str());


	const Aurora::LocString &description = erf.getDescription();
	if (description.getString().empty() && (description.getID() == Aurora::kStrRefInvalid))
		return;

	std::printf("\nDescription:\n");
	std::printf("String reference ID: %u\n", description.getID());

	std::vector<Aurora::LocString::SubLocString> str;
	description.getStrings(str);

	for (std::vector<Aurora::LocString::SubLocString>::iterator s = str.begin(); s != str.end(); ++s) {
		std::printf("\n.=== Description in language %u: ===\n", s->language);
		std::printf("%s\n", s->str.c_str());
		std::printf("'=== ===\n");
	}
}
